From 5fbab772465ca3c1d57b01cbb294be9887da9789 Mon Sep 17 00:00:00 2001
From: zhang xin <xing.z.zhang@intel.com>
Date: Thu, 2 Apr 2009 17:53:34 +0800
Subject: [PATCH] export-pulseaudio-internal-headers-for-module
 This patch export all .h files at pulseaudio_src_path/src/pulse
 pulseaudio_src_path/src/pulsecore pulseaudio_src_path/src/modules
 to $(prefix)/lib/include/pulse-modules-headers/corresponding_subdir_name

You can enable this feature at configure time by
    --enable-install_modules_headers
The default setting is disabled

With this feature you can develop module of pulseaudio without source of PA.

notices:
1. The internal APIs are not stable so best method of developing a module is
in-tree development.
2. You should construct your project by autoconf which provides config.h for internal
headers

Signed-off-by: Vincent Huang<chenglan.huang@intel.com>
---
 configure.ac    |   40 ++++++++++++++++++++++++++++++++
 src/Makefile.am |   68 ++++++++++++++++++++++++++++++++++++++++++++++++++++++-
 2 files changed, 107 insertions(+), 1 deletions(-)

diff -Nur pulseaudio-0.9.15/configure.ac pulseaudio-0.9.15-new/configure.ac
--- pulseaudio-0.9.15/configure.ac	2009-04-14 07:09:53.000000000 +0800
+++ pulseaudio-0.9.15-new/configure.ac	2009-05-06 17:07:17.000000000 +0800
@@ -1199,6 +1199,41 @@
 
 AM_CONDITIONAL([BUILD_MANPAGES], [test "x$manpages" = xyes])
 
+
+AC_ARG_ENABLE([install_modules_headers],
+    AS_HELP_STRING([--enable-install-modules-headers],[optional install modules headers]),
+        [
+            case "${enableval}" in
+                yes) install_modules_headers=yes ;;
+                no) install_modules_headers=no ;;
+                *) AC_MSG_ERROR([bad value ${enableval} for --enable-install-modules-headers]) ;;
+            esac
+        ],
+        [install_modules_headers=no])
+
+if test "x${install_modules_headers}" != xno ; then
+    HAVE_MODULES_HEADERS=1
+	PULSE_HEADERS_PATH="src/pulse"
+	PULSE_HEADERS_LIST=`find "$PULSE_HEADERS_PATH"  -type f | grep -i "^.*\.h$" | sed -e "s:^src/: :g" | tr "\n" ' '`
+	PULSECORE_HEADERS_PATH="src/pulsecore"
+	PULSECORE_HEADERS_LIST=`find "$PULSECORE_HEADERS_PATH"  -type f | grep -i "^.*\.h$" | sed -e "s:^src/: :g" | tr "\n" ' '`
+	PULSEMODULES_HEADERS_PATH="src/modules"
+	PULSEMODULES_HEADERS_LIST=`find "$PULSEMODULES_HEADERS_PATH"  -type f | grep -i "^.*\.h$" | sed -e "s:^src/: :g" | tr "\n" ' '`
+else
+    HAVE_MODULES_HEADERS=0
+fi
+
+if test "x${HAVE_MODULES_HEADERS}" = x1 ; then
+   AC_DEFINE([HAVE_MODULES_HEADERS], 1, [Have install headers?])
+fi
+
+AC_SUBST(PULSE_HEADERS_LIST)
+AC_SUBST(PULSECORE_HEADERS_LIST)
+AC_SUBST(PULSEMODULES_HEADERS_LIST)
+AC_SUBST(HAVE_MODULES_HEADERS)
+AM_CONDITIONAL([HAVE_MODULES_HEADERS], [test "x$HAVE_MODULES_HEADERS" = x1])
+
+
 #### PulseAudio system group & user  #####
 
 AC_ARG_WITH(system_user, AS_HELP_STRING([--with-system-user=<user>],[User for running the PulseAudio daemon as a system-wide instance (pulse)]))
@@ -1425,6 +1460,11 @@
    ENABLE_PER_USER_ESOUND_SOCKET=yes
 fi
 
+ENABLE_MODULES_HEADERS=no
+if test "x${HAVE_MODULES_HEADERS}" = "x1" ; then
+	ENABLE_MODULES_HEADERS=yes
+fi
+
 echo "
  ---{ $PACKAGE_NAME $VERSION }---
 
@@ -1456,7 +1496,7 @@
     Enable PolicyKit:              ${ENABLE_POLKIT}
     Enable IPv6:                   ${ENABLE_IPV6}
     Enable OpenSSL (for Airtunes): ${ENABLE_OPENSSL}
-
+    Enable INSTALL_MODULES_HEADERS:${ENABLE_MODULES_HEADERS}
     System User:                   ${PA_SYSTEM_USER}
     System Group:                  ${PA_SYSTEM_GROUP}
     Realtime Group:                ${PA_REALTIME_GROUP}
diff -Nur pulseaudio-0.9.15/src/Makefile.am pulseaudio-0.9.15-new/src/Makefile.am
--- pulseaudio-0.9.15/src/Makefile.am	2009-03-31 06:31:35.000000000 +0800
+++ pulseaudio-0.9.15-new/src/Makefile.am	2009-05-06 17:10:20.000000000 +0800
@@ -25,7 +25,9 @@
 ###################################
 
 pulseincludedir=$(includedir)/pulse
-pulsecoreincludedir=$(includedir)/pulsecore
+pulseheadersdir=$(includedir)/pulse-modules-headers/pulse
+pulsecoreheadersdir=$(includedir)/pulse-modules-headers/pulsecore
+pulsemodulesheadersdir=$(includedir)/pulse-modules-headers/modules
 pulseconfdir=$(sysconfdir)/pulse
 pulselibexecdir=$(libexecdir)/pulse
 xdgautostartdir=$(sysconfdir)/xdg/autostart
@@ -675,6 +677,12 @@
 		libpulse.la \
 		libpulse-simple.la
 
+if HAVE_MODULES_HEADERS
+pulseinclude_header=$(PULSE_HEADERS_LIST)
+pulsecoreinclude_header=$(PULSECORE_HEADERS_LIST)
+pulsemodulesinclude_header=$(PULSEMODULES_HEADERS_LIST)
+endif
+
 if HAVE_AVAHI
 pulseinclude_HEADERS += \
 		pulse/browser.h
@@ -1650,3 +1658,63 @@
 update-all: update-ffmpeg update-sbc update-map-file
 
 .PHONY: utils/padsp massif update-all update-ffmpeg update-sbc update-map-file
+
+
+if HAVE_MODULES_HEADERS
+install-pulse-headers: $(pulseinclude_header)
+	test -z "$(pulseheadersdir)" || /bin/mkdir -p "$(DESTDIR)$(pulseheadersdir)"
+	@list='$(pulseinclude_header)';for p in $$list; do \
+		if test -f "$$p"; then d=; else d="$(srcdir)/"; fi; \
+		f=`echo "$$p" | sed -e "s:^pulse/::g"`; \
+	    echo "/usr/bin/install -c -D -m 644 '$$d$$p' '$(DESTDIR)$(pulseheadersdir)/$$f'"; \
+		/usr/bin/install -c -D -m 644 "$$d$$p" "$(DESTDIR)$(pulseheadersdir)/$$f"; \
+	done
+
+uninstall-pulse-headers:
+	@list='$(pulseinclude_header)';for p in $$list; do \
+		f=`echo "$$p" | sed -e "s:^pulse/::g"`; \
+		echo " rm -f '$(DESTDIR)$(pulseheadersdir)/$$f'"; \
+		rm -f "$(DESTDIR)$(pulseheadersdir)/$$f"; \
+	done
+
+install-pulsecore-headers: $(pulsecoreinclude_header)
+	test -z "$(pulsecoreheadersdir)" || /bin/mkdir -p "$(DESTDIR)$(pulsecoreheadersdir)"
+	@list='$(pulsecoreinclude_header)';for p in $$list; do \
+		if test -f "$$p"; then d=; else d="$(srcdir)/"; fi; \
+		f=`echo "$$p" | sed -e "s:^pulsecore/::g"`; \
+	    echo "/usr/bin/install -c -D -m 644 '$$d$$p' '$(DESTDIR)$(pulsecoreheadersdir)/$$f'"; \
+		/usr/bin/install -c -D -m 644 "$$d$$p" "$(DESTDIR)$(pulsecoreheadersdir)/$$f"; \
+	done
+
+uninstall-pulsecore-headers:
+	@list='$(pulsecoreinclude_header)';for p in $$list; do \
+		f=`echo "$$p" | sed -e "s:^pulsecore/::g"`; \
+		echo " rm -f '$(DESTDIR)$(pulsecoreheadersdir)/$$f'"; \
+		rm -f "$(DESTDIR)$(pulsecoreheadersdir)/$$f"; \
+	done
+
+install-pulsemodules-headers: $(pulsemodulesinclude_header)
+	test -z "$(pulsemodulesheadersdir)" || /bin/mkdir -p "$(DESTDIR)$(pulsemodulesheadersdir)"
+	@list='$(pulsemodulesinclude_header)';for p in $$list; do \
+		if test -f "$$p"; then d=; else d="$(srcdir)/"; fi; \
+		f=`echo "$$p" | sed -e "s:^modules/::g"`; \
+	    echo "/usr/bin/install -c -D -m 644 '$$d$$p' '$(DESTDIR)$(pulsemodulesheadersdir)/$$f'"; \
+		/usr/bin/install -c -D -m 644 "$$d$$p" "$(DESTDIR)$(pulsemodulesheadersdir)/$$f"; \
+	done
+
+uninstall-pulsemodules-headers:
+	@list='$(pulsemodulesinclude_header)';for p in $$list; do \
+		f=`echo "$$p" | sed -e "s:^modules/::g"`; \
+		echo " rm -f '$(DESTDIR)$(pulsemodulesheadersdir)/$$f'"; \
+		rm -f "$(DESTDIR)$(pulsemodulesheadersdir)/$$f"; \
+	done
+	
+install-data-local: install-pulse-headers  install-pulsecore-headers install-pulsemodules-headers
+
+uninstall-local: uninstall-pulse-headers  uninstall-pulsecore-headers uninstall-pulsemodules-headers
+
+endif
+
+
+
+


